### PR TITLE
Fix - Flaky QueryBuilder tests

### LINKED ISSUE #37184: Flaky QueryBuilder.unit.spec.tsx
https://metaboat.slack.com/archives/C5XHN8GLW/p1702944807217989

https://github.com/metabase/metabase/actions/runs/7255154589/job/19765638188

```
Summary of all failing tests
FAIL frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx (126.407 s)
  ● QueryBuilder › beforeunload events › creating native questions › should trigger beforeunload event when leaving new non-empty native question

thrown: "Caught unmocked POST request to: http://localhost/api/card/1/query"
```

Perhaps related to #34956

----

Old related fixes:
- https://metaboat.slack.com/archives/C057T1QTB3L/p1697183111294489
- #34661

### PR DESCRIPTION
**Pull request by kamilmielnik**
Fixes #37184

### Description

Waiting for all network requests to finish in the `setup` function was crucial to fix this flake.
Calling `waitForLoaderToBeRemoved` was insufficient as not every request has a loader associated with it in the UI.

Fixing required adding typing for `State["requests"]`.

### How to verify

Stress test succeeded (20 runs): https://github.com/metabase/metabase/actions/runs/7960055696

### CODE REVIEW

File: frontend/src/metabase/query_builder/containers/test-utils.tsx [hunk 1]
```diff
@@ -278,7 +283,7 @@ export const setup = async ({
     },
   );
 
-  await waitForLoaderToBeRemoved();
+  await waitForLoadingRequests(getState);
```
Comment (kamilmielnik) [inline frontend/src/metabase/query_builder/containers/test-utils.tsx:286]: This is the fix

File: frontend/src/metabase/query_builder/containers/test-utils.tsx [hunk 2]
```diff
@@ -287,6 +292,25 @@ export const setup = async ({
...
 };
 
+const waitForLoadingRequests = async (getState: () => State) => {
+  await waitFor(
+    () => {
+      const requests = getRequests(getState());
+      const areRequestsLoading = requests.some(request => request.loading);
+      expect(areRequestsLoading).toBe(false);
+    },
+    { timeout: 5000 },
```
Comment (kamilmielnik) [inline frontend/src/metabase/query_builder/containers/test-utils.tsx:302]: The default 1s timeout was not always enough.

**Review by iethree — APPROVED** [review]
👏

**Review by uladzimirdev — APPROVED** [review]
🤞

**Pull request merged**